### PR TITLE
fix: infer missing platform fields of optional dependencies from the package name

### LINKED ISSUE #9940: `supportedArchitecture` don't works as expected
### Verify latest release

- [x] I verified that the issue exists in the latest pnpm release

### pnpm version

10.15.1

### Which area(s) of pnpm are affected? (leave empty if unsure)

_No response_

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/ghoullier/pnpm-supported-architectures-issue

### Reproduction steps

1. clone https://github.com/ghoullier/pnpm-supported-architectures-issue
2. run `pnpm install`
3. run `pnpx node-modules-inspector` OR inspect `./node_modules/.pnpm` folder

### Describe the Bug

pnpm install all the optional dependencies defined in `esbuild`

```json
{
  "optionalDependencies": {
    "@esbuild/aix-ppc64": "0.25.9",
    "@esbuild/android-arm": "0.25.9",
    "@esbuild/android-arm64": "0.25.9",
    "@esbuild/android-x64": "0.25.9",
    "@esbuild/darwin-arm64": "0.25.9",
    "@esbuild/darwin-x64": "0.25.9",
    "@esbuild/freebsd-arm64": "0.25.9",
    "@esbuild/freebsd-x64": "0.25.9",
    "@esbuild/linux-arm": "0.25.9",
    "@esbuild/linux-arm64": "0.25.9",
    "@esbuild/linux-ia32": "0.25.9",
    "@esbuild/linux-loong64": "0.25.9",
    "@esbuild/linux-mips64el": "0.25.9",
    "@esbuild/linux-ppc64": "0.25.9",
    "@esbuild/linux-riscv64": "0.25.9",
    "@esbuild/linux-s390x": "0.25.9",
    "@esbuild/linux-x64": "0.25.9",
    "@esbuild/netbsd-arm64": "0.25.9",
    "@esbuild/netbsd-x64": "0.25.9",
    "@esbuild/openbsd-arm64": "0.25.9",
    "@esbuild/openbsd-x64": "0.25.9",
    "@esbuild/openharmony-arm64": "0.25.9",
    "@esbuild/sunos-x64": "0.25.9",
    "@esbuild/win32-arm64": "0.25.9",
    "@esbuild/win32-ia32": "0.25.9",
    "@esbuild/win32-x64": "0.25.9"
  }
}
```

### Expected Behavior

According to the [pnpm documentation](https://pnpm.io/settings#supportedarchitectures) and given the configuration in `pnpm-workspace.yaml`

```yaml
onlyBuiltDependencies:
  - esbuild

supportedArchitectures:
  cpu: [linux]
  os: [arm64]
```

`pnpm` should only install the following architectures:

- Linux arm64

### Which Node.js version are you using?

v24.6.0

### Which operating systems have you used?

- [x] macOS
- [ ] Windows
- [ ] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

### LINKED ISSUE #11702: pnpm downloads binaries for other platforms even with explicit `supportedArchitectures`
### Verify latest release

- [x] I verified that the issue exists in the latest pnpm release

### pnpm version

11.1.2

### Which area(s) of pnpm are affected? (leave empty if unsure)

_No response_

### Link to the code that reproduces this issue or a replay of the bug

_No response_

### Reproduction steps

- set `supportedArchitectures`
- run `pnpm add -D nx`

### Describe the Bug

even though i explicitly define my spec (os, cpu, and libc)

```yaml
supportedArchitectures:
  os:
    - linux
  cpu:
    - x64
  libc:
    - glibc
```

pnpm still downloads binaries for macos, freebsd and windows

```
Downloading @nx/nx-win32-arm64-msvc@22.7.1: 6.57 MB/6.57 MB, done
Downloading @nx/nx-linux-arm64-musl@22.7.1: 8.64 MB/8.64 MB, done
Downloading @nx/nx-darwin-arm64@22.7.1: 7.72 MB/7.72 MB, done
Downloading @nx/nx-freebsd-x64@22.7.1: 8.43 MB/8.43 MB, done
Downloading @nx/nx-linux-arm-gnueabihf@22.7.1: 8.66 MB/8.66 MB, done
Downloading @nx/nx-darwin-x64@22.7.1: 8.00 MB/8.00 MB, done
Downloading @nx/nx-win32-x64-msvc@22.7.1: 7.00 MB/7.00 MB, done
```

pnpm-lock.yaml
```yaml
  nx@22.7.1:
    dependencies:
       ...
    optionalDependencies:
      '@nx/nx-darwin-arm64': 22.7.1
      '@nx/nx-darwin-x64': 22.7.1
      '@nx/nx-freebsd-x64': 22.7.1
      '@nx/nx-linux-arm-gnueabihf': 22.7.1
      '@nx/nx-linux-arm64-gnu': 22.7.1
      '@nx/nx-linux-arm64-musl': 22.7.1
      '@nx/nx-linux-x64-gnu': 22.7.1
      '@nx/nx-linux-x64-musl': 22.7.1
      '@nx/nx-win32-arm64-msvc': 22.7.1
      '@nx/nx-win32-x64-msvc': 22.7.1
    transitivePeerDependencies:
      - debug

# ...

  '@nx/nx-darwin-arm64@22.7.1':
    resolution: ...
    cpu: [arm64]
    os: [darwin]

  '@nx/nx-darwin-x64@22.7.1':
    resolution: ...
    cpu: [x64]
    os: [darwin]

  '@nx/nx-freebsd-x64@22.7.1':
    resolution: ...
    cpu: [x64]
    os: [freebsd]

  '@nx/nx-linux-arm-gnueabihf@22.7.1':
    resolution: ...
    cpu: [arm]
    os: [linux]

  '@nx/nx-linux-arm64-gnu@22.7.1':
    resolution: ...
    cpu: [arm64]
    os: [linux]
    libc: [glibc]

  '@nx/nx-linux-arm64-musl@22.7.1':
    resolution: ...
    cpu: [arm64]
    os: [linux]
    libc: [musl]

  '@nx/nx-linux-x64-gnu@22.7.1':
    resolution: ...
    cpu: [x64]
    os: [linux]
    libc: [glibc]

  '@nx/nx-linux-x64-musl@22.7.1':
    resolution: ...
    cpu: [x64]
    os: [linux]
    libc: [musl]

  '@nx/nx-win32-arm64-msvc@22.7.1':
    resolution: ...
    cpu: [arm64]
    os: [win32]

  '@nx/nx-win32-x64-msvc@22.7.1':
    resolution: ...
    cpu: [x64]
    os: [win32]
```

I also reproduced this behavior in v10.33.4

Discord thread: https://discord.com/channels/731599538665553971/1503400934464553094/1503808624651276402

### Expected Behavior

it only installs `@nx/nx-linux-x64-gnu`

### Which Node.js version are you using?

24.15.0

### Which operating systems have you used?

- [ ] macOS
- [ ] Windows
- [x] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

Ubuntu 24.10

### PR DESCRIPTION
**Pull request by zkochan**
## Problem

Closes https://github.com/pnpm/pnpm/issues/11702
Closes https://github.com/pnpm/pnpm/issues/9940

Some registries (and registry proxies) strip the `os`/`cpu`/`libc` fields — or just `libc` — from the version objects of the packuments they serve. pnpm's resolution then saw every platform-specific optional dependency (e.g. the `@nx/nx-*`, `@esbuild/*` binaries) as platform-unrestricted, with two consequences:

1. **Every platform's binaries were downloaded and installed**, regardless of `supportedArchitectures` or the current platform.
2. **The lockfile was written without the platform fields** (this is visible in the lockfile committed to the reproduction repo of pnpm/pnpm#9940 — it has `engines` but no `cpu`/`os` on any `@esbuild/*` entry). Such a lockfile then breaks installs *for every machine that uses it*, even ones talking to a healthy registry, because headless installs decide installability from the lockfile's platform fields.

Reproduced end-to-end by putting a metadata-stripping proxy in front of registry.npmjs.org: `pnpm add -D nx` with `supportedArchitectures: {os: [linux], cpu: [x64], libc: [glibc]}` downloaded and installed all 10 `@nx/nx-*` platform binaries and wrote a lockfile with zero `cpu`/`os` fields.

## Fix

Platform-specific binary packages encode their platform in the package name (`@nx/nx-win32-arm64-msvc`, `@esbuild/linux-x64`, `turbo-windows-64`, `bun-linux-aarch64`, …), so `packageIsInstallable` now fills the **missing** platform fields of an *optional* dependency from the name's tokens. Every install path — fresh resolution (package requester), headless install with the isolated linker (`lockfileToDepGraph`), and the hoisted linker (`lockfileToHoistedDepGraph`) — already decides installability through that one check *before* fetching, so foreign-platform binaries are skipped **without downloading them at all**, and existing lockfiles that lack the fields work too.

Inference is deliberately conservative:

- It only runs for optional dependencies, and declared manifest/lockfile fields always take precedence — each field is filled only when missing. In particular, a missing `libc` alone is inferred from the name (`…-linux-x64-musl` vs `…-linux-x64-gnu`), which also covers registries that strip only `libc` (the pnpm/pnpm#7362 / CodeArtifact shape).
- A package that declares no platform fields at all is treated as platform-specific only when an **operating system** token is recognized in its name, so a generic name segment (e.g. `arm` in `is-arm`) never gets a cross-platform package skipped.
- Packages whose names carry no platform tokens (fsevents-shaped) simply keep today's behavior.

## Tests

- `config/package-is-installable`: unit tests for the inference over real-world names (nx, esbuild, biome, turbo, bun, sharp, …) and for the precedence/gating rules, including the libc-only case.
- `package-requester`: a resolver wrapper that strips the platform fields; asserts the package is reported not installable **and no fetch is started**.
- `deps-installer` e2e: a real metadata-stripping HTTP proxy in front of the registry mock; asserts only the matching binary is installed and the foreign ones are not even in the store.
- `deps-installer` e2e: lockfile entries hand-stripped of platform fields + frozen install, for both `isolated` and `hoisted` node linkers.

All of these fail without the fix. Verified end-to-end in Docker against a stripping proxy: `pnpm add -D nx` fetches exactly one binary (`@nx/nx-linux-x64-gnu`), and a frozen install from a lockfile with all platform fields removed materializes only that binary — including the musl/gnu disambiguation, which comes from the name.

## pacquet

Ported in the same PR (`fix(package-is-installable): infer missing platform fields of optional deps from the package name`): `infer_platform_from_package_name` + `inferred_platform` in `pacquet-package-is-installable`, applied inside `package_is_installable` (hoisted linker) and in `compute_skipped_snapshots` (isolated linker, with the check cache keyed by the snapshot's `optional` flag since the verdicts can differ). The `any_installability_constraint` fast path now also considers optional snapshots whose names infer a platform their metadata row doesn't declare, so the inference is reachable on lockfiles without any declared constraint. Unit tests port the same table and gating cases; `installability` pass tests cover the skip, the libc-only disambiguation, the non-optional exemption, and the fast-path interaction.

---
Written by an agent (Claude Code, claude-fable-5).

## Summary by CodeRabbit

* **Bug Fixes**
  * Optional, platform-specific dependencies are now inferred from package names and correctly skipped when incompatible even if registry metadata omits platform fields, preventing unwanted downloads.

* **Tests**
  * Added end-to-end and unit tests covering name-based platform inference, install skipping across link modes, registry/lockfile stripping scenarios, and related edge cases.

* **Chores**
  * Spell-check dictionary updated and a patch-level changeset/version bump added.

### CODE REVIEW

File: installing/package-requester/src/packageRequester.ts [hunk 1]
```diff
@@ -311,9 +312,27 @@ async function resolveAndFetch (
...
+      recheckInstallability = true
+    }
   }
-  // Check installability now that we have the manifest (for git/tarball packages without registry metadata)
-  if (isInstallable === undefined && manifest != null) {
+  // Check installability using the fetched manifest — either the registry
+  // metadata had none (git/tarball packages), or it was missing the
+  // platform fields that the bundled manifest declares.
+  if (recheckInstallability && manifest != null) {
     isInstallable = ctx.force === true || packageIsInstallable(id, manifest, {
```
[the remark this answers is not in the document]
Comment (zkochan) [reply installing/package-requester/src/packageRequester.ts]: Resolved by the rework in 34875b2d7c: the bundled-manifest verification this comment refers to was replaced with name-based platform inference (`inferPlatformFromPackageName`), which runs inside `packageIsInstallable` before the fetch decision. Foreign-platform optional tarballs are no longer fetched at all — the e2e test now asserts `storeHasNot` for them.

---
Written by an agent (Claude Code, claude-fable-5).

File: deps/graph-builder/src/lockfileToDepGraph.ts [hunk 2]
```diff
@@ -283,15 +283,49 @@ async function buildGraphFromPackages (
...
+                nodeVersion: opts.nodeVersion,
+                optional: true,
+                supportedArchitectures: opts.supportedArchitectures,
+              }) === false
+            ) {
+              delete locationByDepPath[depPath]
+              opts.skipped.add(depPath)
+              return
+            }
+          }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply deps/graph-builder/src/lockfileToDepGraph.ts]: Resolved by the rework in 34875b2d7c: the post-fetch verification was removed. The name inference now runs inside `packageIsInstallable`, which `lockfileToDepGraph` calls before any of the present/unchanged fast paths, so there is no path that bypasses the check.

---
Written by an agent (Claude Code, claude-fable-5).

File: installing/deps-restorer/src/lockfileToHoistedDepGraph.ts [hunk 3]
```diff
@@ -245,16 +245,49 @@ async function fetchDeps (
...
+              lockfileDir: opts.lockfileDir,
+              nodeVersion: opts.nodeVersion,
+              optional: true,
+              supportedArchitectures: opts.supportedArchitectures,
+            }) === false
+          ) {
+            opts.skipped.add(depPath)
+            return
+          }
+        }
```
[the remark this answers is not in the document]
Comment (zkochan) [reply installing/deps-restorer/src/lockfileToHoistedDepGraph.ts]: Resolved by the rework in 34875b2d7c: the post-fetch verification was removed. The name inference now runs inside `packageIsInstallable`, which `lockfileToHoistedDepGraph` calls before the `skipFetch` branch, so already-present packages are re-validated on every install like in the isolated linker.

---
Written by an agent (Claude Code, claude-fable-5).

**Comment by zkochan** [comment]
Re CodeRabbit's suggestion to apply the name inference in `pacquet/crates/env-installer/src/install_config_deps.rs`: that gap is intentional. The config-dependencies path deliberately uses `check_package` (no inference), mirroring pnpm's `installConfigDeps`, which calls `checkPackage` rather than `packageIsInstallable` — the env lockfile records all platform variants on purpose and pnpm applies no name inference there either. Per pacquet's cardinal rule, pnpm is the source of truth; if we ever want inference in the config-deps path, it should land in pnpm first.

---
Written by an agent (Claude Code, claude-fable-5).
